prov/sm2/src/sm2.h

```diff
@@ -82,6 +82,15 @@
 #define SM2_ATOMIC_INJECT_SIZE	    (SM2_INJECT_SIZE - sizeof(struct sm2_atomic_hdr))
 #define SM2_ATOMIC_COMP_INJECT_SIZE (SM2_ATOMIC_INJECT_SIZE / 2)
 
+/* protocol flags */
+#define SM2_SAR_LAST_MESSAGE_FLAG (1 << 1)
+#define SM2_SAR_RETURN		  (1 << 2)
+#define SM2_SAR_ERROR_FLAG	  (1 << 3)
+#define SM2_SAR_STATUS_COMPLETED  (1 << 4)
+
```

Comment:
This patch is adding RMA but it looks like it's adding SAR stuff as well? Please separate

Reply:
We intend on introducing SAR for messages and for RMA.  Rather than have one massive PR I've started with implementing RMA using SAR, and intend to do messages with SAR next.  One of them has to come first.  I cannot have a commit introducing SAR without having some mechanism which uses it.

Reply:
I think it's very confusing to have it come in as one. SAR and RMA are two completely different things. SAR is a protocol. RMA is API functionality. SAR will be needed for messages and RMA (the flow should be totally the same with both, just the way you resolve the destination address is different) so doesn't rely on RMA. RMA should also work the same as the other protocols as well so you should be able to use RMA with inline/inject so it shouldn't rely on SAR. SAR and RMA are completely separate. I think avoiding massive PRs is great and I think that's what this is if you combine them. 
I think you need to either add SAR for regular messages first and then add RMA for all protocols or add RMA for the existing protocols and then add SAR for both. Either order is fine with me, but merging them into one is extremely confusing and hard to read and review.

Reply:
> SAR is a protocol. RMA is API functionality.

100% Agree and I understand what you are saying.  However I am not a prescient developer.  I have a concept for a SAR protocol in my head, and I have a concept of how that protocol fits into implementing RMA and messaging functionalities.

It seemed like RMA capability was a good place to start because other developers were providing large message capabilities, and RMA was the sore-thumb of missing functionality.

I did my best to clearly delineate SAR that I expected to re-use in the future for messaging by including SAR in the name, however as I made progress my mental model of which functions would be re-used and how was adjusted, and I'm sure they'll continue to adjust when messages are implemented.  Please don't mistake SAR being used for many RMA operations in this commit to mean that I have somehow confused the two.

I do expect that the Messaging-using-SAR commit will probably further solidify and clarify where the boundaries fall, but given the pace of development, the number of revisions each PR gets, and the amount of parallel work which would be hurt by merge conflicts, I think the choice to implement RMA with SAR and then implement Messages with SAR is the best way forward.  These both need to happen because the SAR-protocol path provides large message when CMA is not enabled and this PR should provide all-size support for RMA, even if in the future we have a faster option.

> you should be able to use RMA with inline/inject so it shouldn't rely on SAR.
The SAR I've written gracefully degrades to something akin to inject, but I'm using the SAR header, so it remains "SAR".

Reply:
Sorry, I wasn't suggesting you were confusing the two. I was saying that putting them together in the same patch is confusing to me as a reviewer because it's hard to tell if a change has to do with SAR or RMA specifically. 
I understand that simultaneous development causes headaches but I think it's for that reason to split it out into easier to handle chunks. Taking the easy way out of writing a lot of code and then just merging it to avoid conflicts so people can rebase I think is problematic for the project as a whole. If simultaneous development within your team is causing issues because you are relying on all the code to go into upstream in order for others to pick it up, I would suggest using someone's branch as a development playground instead (ie have everyone merge to Seth's AWS_sm2 branch and everyone can rebase off that) and then squash commits accordingly. There's no reason to rush things into upstream that aren't fully thought out.
If the SAR you have written gracefully degrades to inject, I would think that it would be easy to plug the RMA you've written into the existing inject protocol. That change, I think, would be simple and could be merged right away. My concern with them together is that SAR is a much more complicated beast since it requires multiple round trips and that requires more thorough review. I would rather remove the RMA specific parts of the code so that we can better view the SAR protocol to understand how it moves data around, whether it's a send or a read/write.

Reply:
I'll do my best to split RMA into a commit with inject only followed by SAR.